crates/assists/src/utils/import_assets.rs

```diff
@@ -0,0 +1,254 @@
+//! Look up accessible paths for items.
+use std::collections::BTreeSet;
+
+use either::Either;
+use hir::{AsAssocItem, AssocItemContainer, ModuleDef, Semantics};
+use ide_db::{imports_locator, RootDatabase};
+use rustc_hash::FxHashSet;
+use syntax::{ast, AstNode, SyntaxNode};
+
+use crate::assist_config::InsertUseConfig;
+
+#[derive(Debug)]
+pub(crate) struct ImportAssets {
+    import_candidate: ImportCandidate,
+    module_with_name_to_import: hir::Module,
+    syntax_under_caret: SyntaxNode,
+}
+
+impl ImportAssets {
+    pub(crate) fn new(ctx: &crate::assist_context::AssistContext) -> Option<Self> {
+        if let Some(path_under_caret) = ctx.find_node_at_offset_with_descend::<ast::Path>() {
+            Self::for_regular_path(path_under_caret, &ctx.sema)
+        } else {
+            Self::for_method_call(ctx.find_node_at_offset_with_descend()?, &ctx.sema)
+        }
+    }
```

Comment:
I think this API is loosing some type-safety. Importing `Path` can't yield `TraitMethod::trait_method`, importing `MethodCall` can't give you anything else. 

I wonder if the could would look better if the public API read roughtly:

```rust
fn import_path(ctx, path: &ast::Path) -> PathImportCandidate { ... }
fn import_trait_for_method(ctx, call: &ast::MethodCallExpr -> TraitImportCandidate { ... }
```

You are now more familiar with the code in question, so I defer to your judgment which option is better :-)

Reply:
Okay, I couldn't really come up with anything that wouldn't create a lot of code duplication i think. I cleaned the module a bit more up and refined the ImportCandidate enum.